src/Features/Core/Portable/SplitOrMergeIfStatements/Consecutive/AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,146 @@
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider<TExpressionSyntax>
+        : AbstractMergeIfStatementsCodeRefactoringProvider<TExpressionSyntax>
+        where TExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool IsElseClauseOfIfStatement(SyntaxNode node, out SyntaxNode ifStatementNode);
+
+        protected abstract bool HasElseClauses(SyntaxNode ifStatementNode);
+
+        protected abstract SyntaxNode MergeIfStatements(
+            SyntaxNode firstIfStatementNode, SyntaxNode secondIfStatementNode, TExpressionSyntax condition);
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+            => new MyCodeAction(createChangedDocument, IfKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken)
+        {
+            return CanBeMergedWithParent(syntaxFacts, ifStatement) ||
+                   await CanBeMergedWithPreviousStatementAsync(document, syntaxFacts, ifStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(
+            SyntaxNode root, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, SyntaxGenerator generator)
+        {
+            var previousIfStatement = IsElseClauseOfIfStatement(ifStatement, out var parentIfStatement)
+                ? parentIfStatement
+                : GetPreviousStatement(syntaxFacts, ifStatement);
+
+            var newCondition = (TExpressionSyntax)generator.LogicalOrExpression(
+                syntaxFacts.GetIfStatementCondition(previousIfStatement),
+                syntaxFacts.GetIfStatementCondition(ifStatement));
+
+            var newIfStatement = MergeIfStatements(previousIfStatement, ifStatement, newCondition);
+
+            var newRoot = root.TrackNodes(previousIfStatement, ifStatement);
+            newRoot = newRoot.ReplaceNode(newRoot.GetCurrentNode(previousIfStatement), newIfStatement.WithAdditionalAnnotations(Formatter.Annotation));
+
+            var currentIfStatement = newRoot.GetCurrentNode(ifStatement);
+            if (currentIfStatement != null)
+            {
+                newRoot = newRoot.RemoveNode(currentIfStatement, SyntaxGenerator.DefaultRemoveOptions);
+            }
+
+            return newRoot;
+        }
+
+        private bool CanBeMergedWithParent(ISyntaxFactsService syntaxFacts, SyntaxNode ifStatement)
+        {
+            return IsElseClauseOfIfStatement(ifStatement, out var parentIfStatement) &&
+                   ContainEquivalentStatements(syntaxFacts, ifStatement, parentIfStatement, out _);
```

Comment:
comments about the equivalancy check.

Reply:
What would you like me to comment about that? It's explained at the top of the class that the body statements have to be equal (and it's also pretty obvious IMO).